benchmarks/benchmarks.py

```diff
@@ -98,3 +98,93 @@ def time_run(self):
     def peakmem_run(self):
         """Benchmark the memory usage of changing a single nested series element."""
         self.run()
+
+
+class NestedFrameAddNested:
+    """Benchmark the NestedFrame.add_nested function"""
+
+    n_base = 100
+    layer_size = 1000
+    base_nf = NestedFrame
+    layer_nf = NestedFrame
+
+    def setup(self):
+        """Set up the benchmark environment"""
+        # use provided seed, "None" acts as if no seed is provided
+        randomstate = np.random.RandomState(seed=1)
+
+        # Generate base data
+        base_data = {"a": randomstate.random(self.n_base), "b": randomstate.random(self.n_base) * 2}
+        self.base_nf = NestedFrame(data=base_data)
+
+        layer_data = {
+            "t": randomstate.random(self.layer_size * self.n_base) * 20,
+            "flux": randomstate.random(self.layer_size * self.n_base) * 100,
+            "band": randomstate.choice(["r", "g"], size=self.layer_size * self.n_base),
+            "index": np.arange(self.layer_size * self.n_base) % self.n_base,
+        }
+        self.layer_nf = NestedFrame(data=layer_data).set_index("index")
+
+    def run(self):
+        """Run the benchmark."""
+        self.base_nf.add_nested(self.layer_nf, "nested")
+
+    def time_run(self):
+        """Benchmark the runtime of adding a nested layer"""
+        self.run()
+
+    def peakmem_run(self):
+        """Benchmark the memory usage of adding a nested layer"""
+        self.run()
+
+
+class NestedFrameReduce:
+    """Benchmark the NestedFrame.reduce function"""
+
+    n_base = 100
+    n_nested = 1000
+    nf = NestedFrame
+
+    def setup(self):
+        """Set up the benchmark environment"""
+        self.nf = datasets.generate_data(self.n_base, self.n_nested)
+
+    def run(self):
+        """Run the benchmark."""
+        self.nf.reduce(np.mean, "nested.flux")
+
+    def time_run(self):
+        """Benchmark the runtime of applying the reduce function"""
+        self.run()
+
+    def peakmem_run(self):
+        """Benchmark the memory usage of applying the reduce function"""
+        self.run()
+
+
+class NestedFrameQuery:
+    """Benchmark the NestedFrame.query function"""
+
+    n_base = 100
+    n_nested = 1000
+    nf = NestedFrame
+
+    def setup(self):
+        """Set up the benchmark environment"""
+        self.nf = datasets.generate_data(self.n_base, self.n_nested)
+
+    def run(self):
+        """Run the benchmark."""
+        # Apply base layer query
+        self.nf = self.nf.query("a > 0.2")
+
+        # Apply nested layer query
+        self.nf = self.nf.query("nested.band == 'g'")
```

Comment:
Would we like to split it into two benchmarks?

Reply:
I think it's fine, as having a benchmark for the top layer query is just timing pandas code.

Reply:
on second thought, for consistency with above I opted to just remove the top layer query.